File: Backend/src/main/java/com/github/francisfire/anavis/models/Donor.java
package com.github.francisfire.anavis.models;

import java.util.Date;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@NoArgsConstructor
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
@Document(collection = "donor")
public class Donor {

	@EqualsAndHashCode.Include
	@Id
	private String mail;
	private String officeMail;
	private boolean canDonate;
	private Date lastDonation;
	private DonorCategory category;
	private String name;
	private String surname;
	private Date birthday;
	private String birthPlace;
	private int leftDonationsInYear;
	private Date firstDonationInYear;

	public Donor(String mail, String officeMail, DonorCategory category) {
		this.mail = mail;
		this.officeMail = officeMail;
		this.category = category;
		this.canDonate = true;
		resetLeftDonationsInYear();
	}
	
	public Donor(String mail, String officeMail, DonorCategory category, String name, String surname, Date birthday, String birthPlace) {
		this.mail = mail;
		this.officeMail = officeMail;
		this.category = category;
		this.canDonate = true;
		this.name = name;
		this.surname = surname;
		this.birthday = birthday;
		this.birthPlace = birthPlace;
		resetLeftDonationsInYear();
	}

	public void resetLeftDonationsInYear() {
		if (leftDonationsInYear == 0) {
			switch (category) {
			case MAN:
			case NONFERTILEWOMAN:
				this.leftDonationsInYear = 4;
				break;
			case FERTILEWOMAN:
				this.leftDonationsInYear = 2;
				break;
			}
		}
	}

	public void setLastDonation(Date lastDonation) {
		if (this.lastDonation == null || lastDonation.after(this.lastDonation)) {
			this.lastDonation = lastDonation;
			switch (category) {
			case MAN:
			case NONFERTILEWOMAN:
				if (leftDonationsInYear == 4) {
					firstDonationInYear = lastDonation;
				}
				break;
			case FERTILEWOMAN:
				if (leftDonationsInYear == 2) {
					firstDonationInYear = lastDonation;
				}
				break;
			}
			leftDonationsInYear--;
		}
	}

	public enum DonorCategory {
		MAN, FERTILEWOMAN, NONFERTILEWOMAN
	}
}

File: Frontend/lib/models/donor.dart
class Donor {
  Donor(String mail, String officeMail, DonorCategory category, String name,
      String surname, String birthday, String birthPlace) {
    this._mail = mail;
    this._officeMail = officeMail;
    this._category = category;
    this._name = name;
    this._surname = surname;
    this._birthday = birthday;
    this._birthPlace = birthPlace;
  }

  Donor.complete(
      String mail,
      String officeMail,
      DonorCategory category,
      String name,
      String surname,
      String birthday,
      String birthPlace,
      bool canDonate,
      String lastDonation,
      int leftDonationsInYear,
      String firstDonationInYear) {
    this._mail = mail;
    this._officeMail = officeMail;
    this._category = category;
    this._name = name;
    this._surname = surname;
    this._birthday = birthday;
    this._birthPlace = birthPlace;
    this._canDonate = canDonate;
    this._lastDonation = lastDonation;
    this._leftDonationsInYear = leftDonationsInYear;
    this._firstDonationInYear = firstDonationInYear;
  }

  String _mail;
  String _officeMail;
  bool _canDonate;
  String _lastDonation;
  DonorCategory _category;
  String _name;
  String _surname;
  String _birthday;
  String _birthPlace;
  int _leftDonationsInYear;
  String _firstDonationInYear;

  String getMail() {
    return _mail;
  }

  String getOfficeMail() {
    return _officeMail;
  }

  bool canDonate() {
    return _canDonate;
  }

  String getLastDonation() {
    return _lastDonation;
  }

  DonorCategory getCategory() {
    return _category;
  }

  String getName() {
    return _name;
  }

  String getSurname() {
    return _surname;
  }

  String getBirthday() {
    return _birthday;
  }

  String getBirthPlace() {
    return _birthPlace;
  }

  int getLeftDonationsInYear() {
    return _leftDonationsInYear;
  }

  String getFirstDonationInYear() {
    return _firstDonationInYear;
  }
}

enum DonorCategory { MAN, FERTILEWOMAN, NONFERTILEWOMAN }
